iceoryx_hoofs/test/moduletests/test_polymorphic_handler.cpp

```diff
@@ -0,0 +1,320 @@
+    auto* ptr = static_cast<Interface*>(&alternateHandler);
+
+    EXPECT_EQ(&handler, ptr);
+}
+
+TEST_F(PolymorphicHandler_test, explicitlySettingToDefaultWorks)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "32e4d808-c848-4bf9-b878-e163ca825539");
+    Handler::set(alternateGuard);
+    auto* prevHandler = Handler::set(defaultGuard);
+
+    auto& handler = Handler::get();
+    auto* ptr = static_cast<Interface*>(&defaultHandler);
+
+    EXPECT_EQ(&handler, ptr);
+    ASSERT_NE(prevHandler, nullptr);
+    EXPECT_EQ(prevHandler->id(), ALTERNATE_ID);
+}
+
+TEST_F(PolymorphicHandler_test, returnValueOfSetPointsToPreviousInstance)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "96447d94-ea27-4d51-8959-12e7752728ae");
+    auto& expectedHandler = Handler::get();
+
+    auto* prevHandler = Handler::set(alternateGuard);
+
+    ASSERT_NE(prevHandler, nullptr);
+    EXPECT_EQ(&expectedHandler, prevHandler);
+    EXPECT_FALSE(prevHandler->isActive());
+}
+
+TEST_F(PolymorphicHandler_test, resetToDefaultWorks)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "ef8a99da-22a6-497e-b2ec-bf72cc3ae943");
+    Handler::set(alternateGuard);
+    auto& prevHandler = Handler::get();
+    EXPECT_EQ(prevHandler.id(), ALTERNATE_ID);
+
+    // note that we have to use reset to set it back to the internal default
+    Handler::reset();
+
+    auto& handler = Handler::get();
+    EXPECT_EQ(handler.id(), DEFAULT_ID);
+    EXPECT_TRUE(handler.isActive());
+
+    EXPECT_FALSE(prevHandler.isActive());
+}
+
+TEST_F(PolymorphicHandler_test, setToCurrentHandlerWorks)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "54e22290-a7b4-4552-a18f-953571381d38");
+
+    // change to alternateHandler
+    Handler::set(alternateGuard);
+    EXPECT_TRUE(alternateHandler.isActive());
+
+    // set to alternateHandler again, should stay active
+    // while this is a useless operation, we cannot forbid it via interface
+    auto* prevHandler = Handler::set(alternateGuard);
+    auto& handler = Handler::get();
+
+    EXPECT_EQ(&handler, prevHandler);
+    EXPECT_EQ(prevHandler, &alternateHandler);
+    EXPECT_TRUE(handler.isActive());
+}
+
+TEST_F(PolymorphicHandler_test, settingAfterFinalizeCallsHook)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "171ac802-01b9-4e08-80a6-6f2defecaf6d");
+
+    auto& handler = Handler::get();
+
+    // reset the handler value to non-zero and check later whether they are set to non-zero as expecteded
+    handler.reset();
+    alternateHandler.reset();
+
+    // note that all following tests will also call the after finalize
+    // hook but we only check if we care whether it was called
+    Handler::finalize();
+    auto* prevHandler = Handler::set(alternateGuard);
+    EXPECT_EQ(prevHandler, nullptr);
+
+    // does the hook set the values to the corresponding arguments?
+    EXPECT_EQ(handler.value, DEFAULT_ID);
+    EXPECT_EQ(alternateHandler.value, ALTERNATE_ID);
+}
+
+TEST_F(PolymorphicHandler_test, resetAfterFinalizeCallsHook)
+{
```

Comment:
@FerdinandSpitzschnueffler 

Your observations are all true. It is a little tricky and was not ideal and hence I changed it now.

`Handler::reset()` is not called in setup anymore (tear down instead) but it could be.

But assume it is called. It is important to note that after a finalize is called in any test, it cannot be undone (it is ... final). The intention of the tests is to verify that reset and set both respect it and call the hook instead of changing the handler.

> Why is `alternateHandler.value` is still 0 and not `ALTERNATE_ID`?

Because we reset it to 0 (in this test) and do not change it with the call of the hook as the `alternateHandler` is not passed to the hook anymore. The previous test tries to set default to alternate, but cannot due to finalize. Therefore the current handler remains default in all tests that follow.

We still should have a check for the `alternateHandler` to verify it was not changed unexpectedly by the hook.

This is an artifact of test execution order and we should try to solve it differently but I am not sure we can. We could do both checks in the same test though to avoid problems with test execution but I changed it to a more elaborate solution which should also work independently of test order unless I missed something.